rascaline-c-api/src/lib.rs

```diff
@@ -8,7 +8,65 @@ mod utils;
 #[macro_use]
 mod status;
 pub use self::status::{catch_unwind, rascal_status_t};
+pub use log::{error, info, warn, Record, Level, Metadata, LevelFilter};
+use std::ffi::CString;
+use std::sync::Mutex;
+use lazy_static::lazy_static;
 
 mod system;
 mod descriptor;
 mod calculator;
+
+pub type RascalLoggingCallback = Option<unsafe extern fn(level: i32, message: *const std::os::raw::c_char)>;
+
+// Mutex cannot use rust static
+// see https://stackoverflow.com/a/27826181
+lazy_static! {
+    static ref GLOBAL_CALLBACK: Mutex<RascalLoggingCallback> = Mutex::new(None);
+}
+
+struct RascalLogger;
+
+#[no_mangle]
+pub unsafe extern fn rascal_set_logging_callback(callback: RascalLoggingCallback, log_level: i32) {
```

Comment:
I'm not sold on exposing the `max_log_level` to the end user directly (in the second parameter here). I think the main use case for this feature is to make sure warnings are visible by the end user; and secondary use case is to give them some info on the running calculation (i.e. `info!` logs). 

Both use cases are well served with simpler code that do not need to deal with the max log level.

Reply:
In any case the end user can just filter the logs in his/her logging utility. So setting the max log level to the maximum level and let the user decide what to filter out, this was the first solution I implemented. But then I was worried that if one has a huge trace log at some point, it could cause performance issues because it has to be forwarded to the logger of the end user before it is decided to not log it. So I thought that the end user should have the possibility to filter out beforehand on the rust side.

Reply:
In the calculator was a comment that crossing the FFI boundary often can cause performance issues, and forwarding a lot of messages seemed to me like a possible problem in future. We could also enforce that the verbose messages are within a `debug_assertion` block, but then I don't know how well this works with other rust libraries. So assume we import a crate which uses !debug not in `debug_assertion` blocks, then we again have the original problem. I am not sure if this are really concerns which could happen in practice.

Reply:
About performance, I don't forsee us sending 10 000 log messages per structures, but there can easily be this many neighbors. So performance might not be an issue for logging =)

About having a lot of debug/trace messages coming from other crates, we could set the max log level to `info` when registering the boxed logger. 

We can always revisit this decision if logging ends up being a performance issue down the line. But in general I find that not trying to optimize until we need it result in cleaner and easier to understand code.

Reply:
But for debugging purpose you want to be able to set it to a higher log levels. Also  there might be case where you want a log message on a per structure level for debugging, of course while you are debugging you would only run it on one structure, but then you need to control the log level.
What do you think about: Set it by default to info, as you said, with an additional `debug_assertion` which sets it to debug level if compiled in debug mode?

Reply:
Using `cfg=debug_assertions` for this sounds good! 

The main tool I use for temporary debugging is the [dbg!](https://doc.rust-lang.org/std/macro.dbg.html) macro, which sends output to the standard error anyway, and is not passed to Python. With this PR merged, I can see myself adding temporary `info!` or `debug!` calls instead =)